Framework/HistogramData/inc/MantidHistogramData/FixedLengthVector.h

```diff
@@ -0,0 +1,124 @@
+#ifndef MANTID_HISTOGRAMDATA_FIXEDLENGTHVECTOR_H_
+#define MANTID_HISTOGRAMDATA_FIXEDLENGTHVECTOR_H_
+
+#include "MantidHistogramData/DllConfig.h"
+
+#include <stdexcept>
+#include <vector>
+
+namespace Mantid {
+namespace HistogramData {
+
+class Histogram;
+
+namespace detail {
+
+/** FixedLengthVector
+
+  Base class providing an object similar to std::vector, with the key difference
+  that the length cannot be changed after creation. This is an implementation
+  detail of HistogramData::Histogram, HistogramData::BinEdges, and
+  HistogramData::Points and is not intended for direct use in client code.
+
+  Copyright &copy; 2016 ISIS Rutherford Appleton Laboratory, NScD Oak Ridge
+  National Laboratory & European Spallation Source
+
+  This file is part of Mantid.
+
+  Mantid is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 3 of the License, or
+  (at your option) any later version.
+
+  Mantid is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+template <class T> class FixedLengthVector {
```

Comment:
`FixedLengthVector` is implemented using the curiously recurring template pattern (CRTP). `T` is the type of the derived type. It is not directly used in `FixedLengthVector`, but it makes derived types completely distinct. For example, we will later introduce `HistogramY`, inheriting from `FixedLengthVector<HistogramY>`. As a consequence, mixing `HistogramX` and `HistogramY` is simply impossible.

I think this more or less answers also your question regarding `VMDBase`. Furthermore `VMDBase` seems to be specificially for short vectors, whereas `FixedLengthVector` is for arbitrary length.
Additionally, I am using the CRTP to have something like "mixins". For example, `class Scalable` can be used to provide scaling for any class inheriting from `FixedLengthVector`. `Scalable` works only if there are iterators available. Later I want to add `class Addable`, to make `FixedLengthVectors` addable. However, `HistogramX` is not `Addable` (adding bin edges would not make sense), only `HistogramY` is. That is, the `operator+` cannot be implemented in the base type. Similarely other operations in `VMDBase` are not meaningful for `HistogramX` or `HistogramY`.

So, I don't think it makes sense to merge `VMDBase` and `FixedLengthVector`. Rather, `VMDBase` might be a candidate where the `Eigen` library would be useful. @MichaelWedel had a look at this recently.